meeteval/wer/wer/time_constrained_orc.py

```diff
@@ -0,0 +1,145 @@
+import meeteval
+from meeteval.wer import ErrorRate, combine_error_rates
+from meeteval.wer.wer.orc import OrcErrorRate
+from meeteval.wer.wer.utils import check_single_filename
+
+import typing
+
+if typing.TYPE_CHECKING:
+    from meeteval.io import STM
+    from meeteval.wer.wer.cp import CPErrorRate
+
+__all__ = [
+    'time_constrained_orc_wer',
+    'time_constrained_orc_wer_multifile',
+]
+
+
+def time_constrained_orc_wer(
+        reference,
+        hypothesis,
+        collar=0,
+        reference_pseudo_word_level_timing='character_based',
+        hypothesis_pseudo_word_level_timing='character_based_points',
+        hypothesis_sort='segment',
+        reference_sort='segment',
+):
+    """
+    The time-constrained version of the ORC-WER (tcORC-WER).
+
+    Special cases where the reference or hypothesis is empty
+    >>> time_constrained_orc_wer([], [])
+    OrcErrorRate(errors=0, length=0, insertions=0, deletions=0, substitutions=0, assignment=())
+    >>> time_constrained_orc_wer([], [{'session_id': 'a', 'start_time': 0, 'end_time': 1, 'words': 'a', 'speaker': 'A'}])
+    OrcErrorRate(errors=1, length=0, insertions=1, deletions=0, substitutions=0, assignment=())
+    >>> time_constrained_orc_wer([{'session_id': 'a', 'start_time': 0, 'end_time': 1, 'words': 'a', 'speaker': 'A'}], [])
+    OrcErrorRate(error_rate=1.0, errors=1, length=1, insertions=0, deletions=1, substitutions=0, assignment=())
+    """
+    if reference_sort == 'word':
+        raise ValueError(
+            'reference_sort="word" is not supported for time-constrained '
+            'ORC-WER because the ORC-WER assumes that an utterance appears'
+            'continuously in the reference.'
+        )
+
+    # Convert to seglst
+    reference = meeteval.io.asseglst(reference)
+    hypothesis = meeteval.io.asseglst(hypothesis)
+    check_single_filename(reference, hypothesis)
+
+    # Remove empty segments
+    reference = reference.filter(lambda s: s['words'] != '')
+    hypothesis = hypothesis.filter(lambda s: s['words'] != '')
+
+    # Add a segment index to the reference so that we can later find words that
+    # come from the same segment
+    for i, s in enumerate(reference):
+        s['segment_index'] = i
```

Comment:
Hmm. Should the assignment be before the filter?

Reply:
It shouldn't make a difference because the segment index is here just used for grouping and not for sorting. But it could be easier to understand what's happening when moved before the filter operation